### PR TITLE
cmd/scollector: Add dotnet.* metrics for tracking appdomains 

### PR DESCRIPTION
**Pull request by gbrayut**

### CODE REVIEW

**Comment by gbrayut** [comment]
This uses process_dotnet settings from the config file. Proposed default conf file for Stack Exchange:

> process_dotnet=^w3wp
> process_dotnet=^LogStasher
> process_dotnet=^SocketServer
> process_dotnet=^Bonfire.Server
> process_dotnet=^StackServer
> process_dotnet=^MobilePushServer
> process_dotnet=^Scheduler

**Comment by captncraig** [comment]
:thumbsup:

**Comment by gbrayut** [comment]
I removed regexes from function signature to fix name issue. Also renamed regexes to regexesProcesses to prevent collisions in collectors package

File: cmd/scollector/collectors/dotnet_windows.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	WatchProcessesDotNet = func() {
+		if len(regexesDotNet) == 0 {
+			// if no processesDotNet configured in config file, use this set instead.
+			regexesDotNet = append(regexesDotNet, regexp.MustCompile("^w3wp"))
+		}
+		c := &IntervalCollector{
+			F:        c_dotnet_loading,
+			Interval: time.Second * 60,
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: Why not default freq?

File: cmd/scollector/collectors/dotnet_windows.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		}
+		c.init = wmiInit(c, func() interface{} { return &[]Win32_PerfRawData_NETFramework_NETCLRLoading{} }, ``, &dotnetLoadingQuery)
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: Backticks are generally only used when you need "raw string" behavior, just use double quotes. Also I would make the anon func span multiple lines

File: cmd/scollector/collectors/dotnet_windows.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	err := queryWmi(dotnetLoadingQuery, &dst)
+	if err != nil {
+		return nil, err
+	}
+	var md opentsdb.MultiDataPoint
+	for _, v := range dst {
+		if !nameMatches(v.Name, regexesDotNet) {
+			continue
+		}
+
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: No spaces inside a func

File: cmd/scollector/collectors/dotnet_windows.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+func init() {
+	AddProcessDotNetConfig = func(line string) error {
+		reg, err := regexp.Compile(line)
+		if err != nil {
+			return err
+		}
+		regexesDotNet = append(regexesDotNet, reg)
+		return nil
+	}
+
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: No empty lines inside a func

File: cmd/scollector/collectors/collectors.go [hunk 5]
```diff
@@ -67,6 +68,9 @@ var (
 	timestamp = time.Now().Unix()
 	tlock     sync.Mutex
 	AddTags   opentsdb.TagSet
+
+	AddProcessDotNetConfig = func(line string) error { return fmt.Errorf("process_dotnet watching not implemented on this platform") }
```
Comment (kylebrandt) [inline cmd/scollector/collectors/collectors.go]: Span multiple lines

File: cmd/scollector/collectors/processes_windows.go [hunk 6]
```diff
@@ -10,31 +10,28 @@ import (
 	"bosun.org/opentsdb"
 )
 
-var regexes = []*regexp.Regexp{}
+var regexesProcesses = []*regexp.Regexp{}
 
 func AddProcessConfig(line string) error {
 	reg, err := regexp.Compile(line)
```
Comment (kylebrandt) [inline cmd/scollector/collectors/processes_windows.go]: This was existing code, but I think it should Panic if something in the config isn't a valid regex. You can use `regexp.MustCompile` to do this

File: cmd/scollector/collectors/dotnet_windows.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+)
+
+var regexesDotNet = []*regexp.Regexp{}
+
+func init() {
+	AddProcessDotNetConfig = func(line string) error {
+		reg, err := regexp.Compile(line)
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: Same here with MustCompile

File: cmd/scollector/collectors/processes_windows.go (hunk 6 above)
Comment (captncraig) [inline cmd/scollector/collectors/processes_windows.go]: I disagree. Panics should be reserved for truly unexpected things. Someone giving bad input is not unexpected. We should print an error and exit, as we are here. 

MustCompile is best reserved for hardcoded regexes in code that are compiled at init time (before we even look at the config file)

Comment (mjibson) [inline cmd/scollector/collectors/processes_windows.go]: I agree with @captncraig

Comment (Dieterbe) [inline cmd/scollector/collectors/processes_windows.go]: +1 panic only for worst case. friendly errors otherwise

Comment (kylebrandt) [inline cmd/scollector/collectors/processes_windows.go]: I agree with your agreement as long as the program exits when it encounters a regex compile error and doesn't just log it.

**Comment by gbrayut** [comment]
@kylebrandt updated with fixes for CR and current WIP for memory metrics. Should be able to finish it up tomorrow.

**Comment by gbrayut** [comment]
This should be ready for review with Memory stats now. I'll squash before merge but it is easier to see the changes this way.

File: cmd/scollector/collectors/dotnet_windows.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+		if err != nil {
+			return err
+		}
+		regexesDotNet = append(regexesDotNet, reg)
+		return nil
+	}
+	WatchProcessesDotNet = func() {
+		if len(regexesDotNet) == 0 {
+			// if no processesDotNet configured in config file, use this set instead.
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: Capital "If"

File: cmd/scollector/collectors/dotnet_windows.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+			name = raw_name[0]
+			if len(raw_name) == 2 {
+				id = raw_name[1]
+			}
+			// If you have a hash sign in your process name you don't deserve monitoring ;-)
+			if len(raw_name) > 2 {
+				continue
+			}
+		}
+		// A Service match could "overwrite" a process match, but that is probably what we would want
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: Needs period at end

File: cmd/scollector/collectors/dotnet_windows.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+		for _, svc := range svc_dst {
+			if nameMatches(svc.Name, regexesDotNet) {
+				// It is possible the pid has gone and been reused, but I think this unlikely
+				// And I'm not aware of an atomic join we could do anyways
```
Comment (kylebrandt) [inline cmd/scollector/collectors/dotnet_windows.go]: lowercase and, period at end

File: cmd/scollector/collectors/processes_windows.go [hunk 11]
```diff
@@ -10,31 +10,28 @@ import (
...
+	regexesProcesses = append(regexesProcesses, reg)
 	return nil
 }
 
 func WatchProcesses() {
-	if len(regexes) == 0 {
+	if len(regexesProcesses) == 0 {
 		// if no processes configured in config file, use this set instead.
-		regexes = append(regexes, regexp.MustCompile("chrome|powershell|scollector|SocketServer|WinRM|MSSQLSERVER|StackServerProd|StackServerDev|LogStasher"))
+		regexesProcesses = append(regexesProcesses, regexp.MustCompile("chrome|powershell|scollector|SocketServer|WinRM|MSSQLSERVER|StackServerProd|StackServerDev|LogStasher"))
```
Comment (kylebrandt) [inline cmd/scollector/collectors/processes_windows.go]: Get rid of anything SE Specific

Comment (gbrayut) [inline cmd/scollector/collectors/processes_windows.go]: That is is scheduled for after the PR is merged and conf files are deployed. Once we are working as expected on conf files I'll update the default settings to remove anything from stackexchange.

**Comment by kylebrandt** [comment]
You can added "fixes #764" to your commit, when it gets merged, that issue will close automagically.

**Comment by mjibson** [comment]
LGTM

**Pull request merged**